contracts/masset/Masset.sol

```diff
@@ -249,41 +266,65 @@ contract Masset is
+     * in the cache, then just transfer, otherwise reset the cache to the desired mid level by
+     * depositing the delta in the platform
+     */
     function _depositTokens(
         address _bAsset,
         uint256 _bAssetRatio,
         address _integrator,
-        bool _erc20TransferFeeCharged,
-        uint256 _quantity
+        bool _hasTxFee,
+        uint256 _quantity,
+        uint256 _maxCache
     )
         internal
         returns (uint256 quantityDeposited, uint256 ratioedDeposit)
     {
-        quantityDeposited = _depositTokens(_bAsset, _integrator, _erc20TransferFeeCharged, _quantity);
-        ratioedDeposit = quantityDeposited.mulRatioTruncate(_bAssetRatio);
-    }
+        console.log("depositTokens: start");
+        // 1 - Send all to PI, using the opportunity to get the cache balance and net amount transferred
+        (uint256 transferred, uint256 cacheBal) = MassetHelpers.transferReturnBalance(msg.sender, _integrator, _bAsset, _quantity);
+
+        // 2 - Deposit X if necessary
+        //   - Can account here for non-lending market integrated bAssets by simply checking for
+        //     integrator == address(0) || address(this) and then keeping entirely in cache
+        // 2.1 - Deposit if xfer fees
+        if(_hasTxFee){
+            console.log("_depositTokens: hasTxFee");
+            uint256 deposited = IPlatformIntegration(_integrator).deposit(_bAsset, transferred, true);
+            quantityDeposited = StableMath.min(deposited, _quantity);
+        }
+        // 2.2 - Else Deposit X if Cache > %
+        else {
+            // This check is in place to ensure that any token with a txFee is rejected
+            // Audit notes: Assumption made that if no fee is collected here then there is no txfee
+            require(transferred == _quantity, "Asset not fully transferred");
 
-    /** @dev Deposits tokens into the platform integration and returns the deposited amount */
-    function _depositTokens(
-        address _bAsset,
-        address _integrator,
-        bool _erc20TransferFeeCharged,
-        uint256 _quantity
-    )
-        internal
-        returns (uint256 quantityDeposited)
-    {
-        uint256 quantityTransferred = MassetHelpers.transferTokens(msg.sender, _integrator, _bAsset, _erc20TransferFeeCharged, _quantity);
-        uint256 deposited = IPlatformIntegration(_integrator).deposit(_bAsset, quantityTransferred, _erc20TransferFeeCharged);
-        quantityDeposited = StableMath.min(deposited, _quantity);
-    }
+            quantityDeposited = transferred;
+
+            uint256 relativeMaxCache = _maxCache.divRatioPrecisely(_bAssetRatio);
 
+            console.log("_depositTokens: cacheBal: %s vs relativeMaxCache: %s", cacheBal, relativeMaxCache);
+            if(cacheBal >= relativeMaxCache){
+                uint256 delta = cacheBal.sub(relativeMaxCache.div(2));
+                IPlatformIntegration(_integrator).deposit(_bAsset, delta, false);
```

Comment:
Maybe this should only deposit if `0 < delta` since the integration will fail otherwise. Maybe `cacheBal >= relativeMaxCache` could be changed to `cacheBal > relativeMaxCache`.

Reply:
I don't see a scenario in which the delta is == 0

Reply:
Yeah, I think the only scenario I see right now is if both `cacheBal` and `relativeMaxCache` are 0. I think `cacheBal` should always be positive though.

I think the change I'm proposing would simplify the reasoning here. First, it should be easy to see that `delta` is always positive. Second, to me it's more intuitive to think of `cacheBal == relativeMaxCache` as a valid state of the cache since the balance hasn't exceeded the maximum.

Reply:
Agreed, changed.